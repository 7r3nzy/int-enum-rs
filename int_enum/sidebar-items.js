initSidebarItems({"struct":[["IntEnumError","An error when attempting to convert an integer into an enum."]],"trait":[["IntEnum","Trait used for implementations of integer and enum conversions."]]});