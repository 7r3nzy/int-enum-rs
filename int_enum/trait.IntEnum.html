<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `IntEnum` trait in crate `int_enum`."><meta name="keywords" content="rust, rustlang, rust-lang, IntEnum"><title>int_enum::IntEnum - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../dark.css"><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script src="../storage.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="shortcut icon" href="../favicon.ico"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc trait"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../int_enum/index.html'><div class='logo-container'><img src='../rust-logo.png' alt='logo'></div></a><p class='location'>Trait IntEnum</p><div class="sidebar-elems"><div class="block items"><a class="sidebar-title" href="#associated-types">Associated Types</a><div class="sidebar-links"><a href="#associatedtype.Int">Int</a></div><a class="sidebar-title" href="#required-methods">Required Methods</a><div class="sidebar-links"><a href="#tymethod.from_int">from_int</a><a href="#tymethod.int_value">int_value</a></div><a class="sidebar-title" href="#implementors">Implementors</a></div><p class='location'><a href='index.html'>int_enum</a></p><script>window.sidebarCurrent = {name: 'IntEnum', ty: 'trait', relpath: ''};</script><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!"><img src="../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices"></div></div><script src="../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><a id="settings-menu" href="../settings.html"><img src="../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class='fqn'><span class='out-of-band'><span id='render-detail'><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class='inner'>&#x2212;</span>]</a></span><a class='srclink' href='../src/int_enum/lib.rs.html#106-117' title='goto source code'>[src]</a></span><span class='in-band'>Trait <a href='index.html'>int_enum</a>::<wbr><a class="trait" href=''>IntEnum</a></span></h1><div class="docblock type-decl hidden-by-usual-hider"><pre class='rust trait'>pub trait IntEnum: <a class="trait" href="https://doc.rust-lang.org/nightly/core/marker/trait.Copy.html" title="trait core::marker::Copy">Copy</a> {
    type <a href='#associatedtype.Int' class="type">Int</a>: PrimInt;
    fn <a href='#tymethod.int_value' class='fnname'>int_value</a>(self) -&gt; Self::<a class="type" href="../int_enum/trait.IntEnum.html#associatedtype.Int" title="type int_enum::IntEnum::Int">Int</a>;
<div class='item-spacer'></div>    fn <a href='#tymethod.from_int' class='fnname'>from_int</a>(n: Self::<a class="type" href="../int_enum/trait.IntEnum.html#associatedtype.Int" title="type int_enum::IntEnum::Int">Int</a>) -&gt; <a class="enum" href="https://doc.rust-lang.org/nightly/core/result/enum.Result.html" title="enum core::result::Result">Result</a>&lt;Self, <a class="struct" href="../int_enum/struct.IntEnumError.html" title="struct int_enum::IntEnumError">IntEnumError</a>&lt;Self&gt;&gt;<br>&nbsp;&nbsp;&nbsp; <span class="where">where<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Self: <a class="trait" href="https://doc.rust-lang.org/nightly/core/marker/trait.Sized.html" title="trait core::marker::Sized">Sized</a></span>;
}</pre></div><div class='docblock'><p>Trait used for implementations of integer and enum conversions.</p>
<h1 id="examples" class="section-header"><a href="#examples">Examples</a></h1>
<p>Basic usage:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="kw">use</span> <span class="ident">int_enum</span>::<span class="ident">IntEnum</span>;

<span class="attribute">#[<span class="ident">repr</span>(<span class="ident">u8</span>)]</span>
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Clone</span>, <span class="ident">Copy</span>, <span class="ident">Debug</span>, <span class="ident">Eq</span>, <span class="ident">PartialEq</span>, <span class="ident">IntEnum</span>)]</span>
<span class="kw">pub</span> <span class="kw">enum</span> <span class="ident">SmallInt</span> {
    <span class="ident">One</span> <span class="op">=</span> <span class="number">1</span>,
    <span class="ident">Two</span> <span class="op">=</span> <span class="number">2</span>,
}

<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="number">1</span>, <span class="ident">SmallInt</span>::<span class="ident">One</span>.<span class="ident">int_value</span>());
<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="number">2</span>, <span class="ident">SmallInt</span>::<span class="ident">Two</span>.<span class="ident">int_value</span>());

<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="ident">SmallInt</span>::<span class="ident">One</span>, <span class="ident">SmallInt</span>::<span class="ident">from_int</span>(<span class="number">1</span>)<span class="question-mark">?</span>);
<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="ident">SmallInt</span>::<span class="ident">Two</span>, <span class="ident">SmallInt</span>::<span class="ident">from_int</span>(<span class="number">2</span>)<span class="question-mark">?</span>);

<span class="macro">assert</span><span class="macro">!</span>(<span class="ident">SmallInt</span>::<span class="ident">from_int</span>(<span class="number">5</span>).<span class="ident">is_err</span>());</pre></div>
<p>Serde support (requires feature <code>serialize</code>):</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="kw">use</span> <span class="ident">int_enum</span>::<span class="ident">IntEnum</span>;

<span class="attribute">#[<span class="ident">repr</span>(<span class="ident">i8</span>)]</span>
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Clone</span>, <span class="ident">Copy</span>, <span class="ident">Debug</span>, <span class="ident">Eq</span>, <span class="ident">PartialEq</span>, <span class="ident">IntEnum</span>)]</span>
<span class="kw">enum</span> <span class="ident">Num</span> {
    <span class="ident">MinusThree</span> <span class="op">=</span> <span class="op">-</span><span class="number">3</span>,
    <span class="ident">Zero</span> <span class="op">=</span> <span class="number">0</span>,
    <span class="ident">Five</span> <span class="op">=</span> <span class="number">5</span>,
}

<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="string">&quot;-3&quot;</span>, <span class="ident">serde_json</span>::<span class="ident">to_string</span>(<span class="kw-2">&amp;</span><span class="ident">Num</span>::<span class="ident">MinusThree</span>)<span class="question-mark">?</span>);
<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="string">&quot;0&quot;</span>, <span class="ident">serde_json</span>::<span class="ident">to_string</span>(<span class="kw-2">&amp;</span><span class="ident">Num</span>::<span class="ident">Zero</span>)<span class="question-mark">?</span>);
<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="string">&quot;5&quot;</span>, <span class="ident">serde_json</span>::<span class="ident">to_string</span>(<span class="kw-2">&amp;</span><span class="ident">Num</span>::<span class="ident">Five</span>)<span class="question-mark">?</span>);

<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="ident">Num</span>::<span class="ident">MinusThree</span>, <span class="ident">serde_json</span>::<span class="ident">from_str</span>(<span class="string">&quot;-3&quot;</span>)<span class="question-mark">?</span>);
<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="ident">Num</span>::<span class="ident">Zero</span>, <span class="ident">serde_json</span>::<span class="ident">from_str</span>(<span class="string">&quot;0&quot;</span>)<span class="question-mark">?</span>);
<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="ident">Num</span>::<span class="ident">Five</span>, <span class="ident">serde_json</span>::<span class="ident">from_str</span>(<span class="string">&quot;5&quot;</span>)<span class="question-mark">?</span>);

<span class="macro">assert</span><span class="macro">!</span>(<span class="ident">serde_json</span>::<span class="ident">from_str</span>::<span class="op">&lt;</span><span class="ident">Num</span><span class="op">&gt;</span>(<span class="string">&quot;4&quot;</span>).<span class="ident">is_err</span>());</pre></div>
<p><code>From</code> and <code>TryFrom</code> support (requires <code>convert</code> feature):</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="kw">use</span> <span class="ident">std</span>::<span class="ident">convert</span>::<span class="ident">TryFrom</span>;
<span class="kw">use</span> <span class="ident">int_enum</span>::<span class="ident">IntEnum</span>;

<span class="attribute">#[<span class="ident">repr</span>(<span class="ident">u16</span>)]</span>
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Clone</span>, <span class="ident">Copy</span>, <span class="ident">Debug</span>, <span class="ident">Eq</span>, <span class="ident">PartialEq</span>, <span class="ident">IntEnum</span>)]</span>
<span class="kw">enum</span> <span class="ident">Value</span> {
    <span class="ident">A</span> <span class="op">=</span> <span class="number">1_000</span>,
    <span class="ident">B</span> <span class="op">=</span> <span class="number">1_001</span>,
    <span class="ident">C</span> <span class="op">=</span> <span class="number">1_002</span>,
}

<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="number">1_000</span>, <span class="ident">u16</span>::<span class="ident">from</span>(<span class="ident">Value</span>::<span class="ident">A</span>));
<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="number">1_001</span>, <span class="ident">u16</span>::<span class="ident">from</span>(<span class="ident">Value</span>::<span class="ident">B</span>));
<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="number">1_002</span>, <span class="ident">u16</span>::<span class="ident">from</span>(<span class="ident">Value</span>::<span class="ident">C</span>));

<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="ident">Value</span>::<span class="ident">A</span>, <span class="ident">Value</span>::<span class="ident">try_from</span>(<span class="number">1_000</span>)<span class="question-mark">?</span>);
<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="ident">Value</span>::<span class="ident">B</span>, <span class="ident">Value</span>::<span class="ident">try_from</span>(<span class="number">1_001</span>)<span class="question-mark">?</span>);
<span class="macro">assert_eq</span><span class="macro">!</span>(<span class="ident">Value</span>::<span class="ident">C</span>, <span class="ident">Value</span>::<span class="ident">try_from</span>(<span class="number">1_002</span>)<span class="question-mark">?</span>);

<span class="macro">assert</span><span class="macro">!</span>(<span class="ident">Value</span>::<span class="ident">try_from</span>(<span class="number">2_000</span>).<span class="ident">is_err</span>());</pre></div>
</div>
            <h2 id='associated-types' class='small-section-header'>Associated Types<a href='#associated-types' class='anchor'></a></h2><div class='methods'><h3 id='associatedtype.Int' class='method'><code id='Int.t'>type <a href='#associatedtype.Int' class="type">Int</a>: PrimInt</code></h3><div class='docblock'><p>Primitive integer type for conversions.</p>
</div></div><span class='loading-content'>Loading content...</span>
            <h2 id='required-methods' class='small-section-header'>Required methods<a href='#required-methods' class='anchor'></a></h2><div class='methods'><h3 id='tymethod.int_value' class='method'><code id='int_value.v'>fn <a href='#tymethod.int_value' class='fnname'>int_value</a>(self) -&gt; Self::<a class="type" href="../int_enum/trait.IntEnum.html#associatedtype.Int" title="type int_enum::IntEnum::Int">Int</a></code></h3><div class='docblock'><p>Returns the integer value of the enum.</p>
</div><h3 id='tymethod.from_int' class='method'><code id='from_int.v'>fn <a href='#tymethod.from_int' class='fnname'>from_int</a>(n: Self::<a class="type" href="../int_enum/trait.IntEnum.html#associatedtype.Int" title="type int_enum::IntEnum::Int">Int</a>) -&gt; <a class="enum" href="https://doc.rust-lang.org/nightly/core/result/enum.Result.html" title="enum core::result::Result">Result</a>&lt;Self, <a class="struct" href="../int_enum/struct.IntEnumError.html" title="struct int_enum::IntEnumError">IntEnumError</a>&lt;Self&gt;&gt; <span class="where fmt-newline">where<br>&nbsp;&nbsp;&nbsp;&nbsp;Self: <a class="trait" href="https://doc.rust-lang.org/nightly/core/marker/trait.Sized.html" title="trait core::marker::Sized">Sized</a>,&nbsp;</span></code></h3><div class='docblock'><p>Attempts to convert an integer into the enum.</p>
</div></div><span class='loading-content'>Loading content...</span>
            <h2 id='implementors' class='small-section-header'>Implementors<a href='#implementors' class='anchor'></a></h2><div class='item-list' id='implementors-list'></div><span class='loading-content'>Loading content...</span><script type="text/javascript">window.inlined_types=new Set([]);</script><script type="text/javascript" async
                         src="../implementors/int_enum/trait.IntEnum.js">
                 </script></section><section id="search" class="content hidden"></section><section class="footer"></section><script>window.rootPath = "../";window.currentCrate = "int_enum";</script><script src="../aliases.js"></script><script src="../main.js"></script><script defer src="../search-index.js"></script></body></html>