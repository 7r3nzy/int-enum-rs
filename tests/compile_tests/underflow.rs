#![deny(overflowing_literals)]

use int_enum::IntEnum;

#[repr(i8)]
#[derive(Clone, Copy, IntEnum)]
pub enum Underflow {
    UnderMin = -129,
    Min = -128,
}

fn main() {}
