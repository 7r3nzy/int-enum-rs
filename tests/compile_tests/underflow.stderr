error: literal out of range for `i8`
 --> $DIR/underflow.rs:8:17
  |
8 |     UnderMin = -129,
  |                 ^^^
  |
note: the lint level is defined here
 --> $DIR/underflow.rs:1:9
  |
1 | #![deny(overflowing_literals)]
  |         ^^^^^^^^^^^^^^^^^^^^
